### ISSUE TITLE
Consider using pyarrow under the hood for fast ASCII reading

### ISSUE THREAD

**Issue by taldcroft** [astropy/astropy#16869]
The fast-ASCII reader in C is difficult to maintain. Thus far it has not required a large amount of work but efforts like #16860 highlight that some work is needed.

This stems from conversation in #12829 and is basically the same idea as that issue. This option has the obvious downside of being a new dependency, although "fast reading" might be considered optional.

The main benefit would be an independently maintained fast-reader which supports fast multi-threaded reading of large tables. The performance benefits are not yet known but @dhomeier said this evaluation should be feasible.

**Comment by pllim** [comment]
Are you proposing pyarrow as a new required or optional dependency?

**Comment by taldcroft** [comment]
It would be an optional but desirable dependency. But this is honestly a bit more pie-in-the-sky right now, I'm just wondering out loud if pyarrow might work for us from a performance and interface perspective.

**Comment by pllim** [comment]
Might be a good candidate for GSoC if we decide to do that again? Sounds like a self-contained exploration project that can be time-boxed. 💭

**Comment by dhomeier** [comment]
The numbers are looking, well, quite spectacular – even with threading disabled a factor of 2 over the fastest other readers, and that is another 6-7 times speedup in the (default) multi-threaded mode on a M1 with 8 performance, 2 economy cores.
That's using the `read_csv` function, which would handle basic formats with custom (single?) separators – hooking our functionality into that API may still require some work.
But even for a (probably fairly common) subset of table formats making this available as a fast_reader option might be well worth it.
![read_e16](https://github.com/user-attachments/assets/18d512a2-81b2-4cbe-9af2-0a5e211d8eee)
![read_f16](https://github.com/user-attachments/assets/7d3e86bc-1df8-4d6a-9925-29ccf132ba3d)
![read_int30](https://github.com/user-attachments/assets/9aa095f2-7cb2-47af-9585-e077befecb5d)

**Comment by taldcroft** [comment]
@dhomeier - wow, and thanks!! Do you have any assessment of numerical accuracy issues for the outlier floating point cases that you worry about?

**Comment by taldcroft** [comment]
Looking into `pyarrow` as a dependency, a mamba or pip install brings long about 25-30 Mb of new packages. Definitely worth it IMO if this brings a factor of 10 in file reading speed. I'm not sure how well this can fit into the current `io.ascii` interface, but that's TBD. 

Another option I have been mulling over is forking `io.ascii` into `io.text` and simplying the interface and code. Starting with `pyarrow` as the base reader would be a good way to start this.

**Comment by pllim** [comment]
For the first 2 plots in https://github.com/astropy/astropy/issues/16869#issuecomment-2307882059 , hard to see but I think "io.ascii Fast-c" is overlapping with "Fast converter," is that right?

How is pyarrow support for "exotic archs"? I guess when someone issues a PR, we will find out...

**Comment by taldcroft** [comment]
@pllim - pyarrow is already an optional dependency for parquet support. Are those tests run on "exotic" archs?

**Comment by taldcroft** [comment]
> For the first 2 plots in [#16869 (comment)](https://github.com/astropy/astropy/issues/16869#issuecomment-2307882059) , hard to see but I think "io.ascii Fast-c" is overlapping with "Fast converter," is that right?

That matches what I see.

**Comment by pllim** [comment]
Re: exotic archs -- Does not look like it.

```
2024-08-26T06:28:06.4907650Z Package versions: 
2024-08-26T06:28:06.4907923Z Numpy: 1.26.4
2024-08-26T06:28:06.4908187Z Scipy: not available
2024-08-26T06:28:06.4908487Z Matplotlib: not available
2024-08-26T06:28:06.4908795Z h5py: not available
2024-08-26T06:28:06.4909088Z Pandas: not available
2024-08-26T06:28:06.4909377Z PyERFA: 2.0.1.4
2024-08-26T06:28:06.4909636Z Cython: 3.0.11
2024-08-26T06:28:06.4909953Z Scikit-image: not available
2024-08-26T06:28:06.4910573Z asdf-astropy: not available
2024-08-26T06:28:06.4910901Z pyarrow: not available
2024-08-26T06:28:06.4911102Z 
2024-08-26T06:28:06.4911249Z Using Astropy options: remote_data: none.
```

**Comment by mhvk** [comment]
Just from the sidelines, but that speed-up is really quite spectacular! Having pyarrow as an optional dependency, replacing the fast reader, would seem to make sense to me -- typical table use probably doesn't require the fast reader in the first place, but when dealing with big ASCII tables, a factor 100 over plain python (and 10 over our current fast reader) would be quite welcome!

I also generally like the idea of outsourcing problems to specialists! (though it is nice to see we still beat `np.loadtxt` by quite a bit.)

p.s.  And presumably our own .ecsv files would really benefit from this too, since they are just .csv with a bigger header.

**Comment by dhomeier** [comment]
> For the first 2 plots in [#16869 (comment)](https://github.com/astropy/astropy/issues/16869#issuecomment-2307882059) , hard to see but I think "io.ascii Fast-c" is overlapping with "Fast converter," is that right?

Correct, speedups for the fast converter are in the range 1.0-1.05 now, meaning we do not really gain much from `xstrtod` at this point except support for alternative exponential notations. np.loadtxt is significantly slower at the full 16-digit precision, but catching up and outperforming the fast reader already for `{:.8f}` format.

**Comment by bsipocz** [comment]
> Are you proposing pyarrow as a new required or optional dependency?

It is already an optional dependency, we use it in the parquet readers

**Comment by bsipocz** [comment]
> @pllim - pyarrow is already an optional dependency for parquet support. Are those tests run on "exotic" archs?

Ahh, OK, I see that you got to the same point. Then to answer the remaining question:

We do skip the tests when pyarrow is not available, which is more than sensible for the parquet use case, as I don't think those are realistic use cases to work with the huge catalogs which dominate the parquet cases.

**Comment by taldcroft** [comment]
I'm planning[^1] to put together a prototype to see how this might look in astropy and make it easier to play around.

[^1]: Noting that is is much easier to *plan* than *do*.
